wsagent/che-core-api-languageserver/src/main/java/org/eclipse/che/api/languageserver/registry/LanguageServerRegistryImpl.java

```diff
@@ -26,99 +25,117 @@
-                                      ServerInitializer initializer) {
+    public LanguageServerRegistryImpl(Set<LanguageServerLauncher> languageServerLaunchers, Set<LanguageDescription> languages,
+                                      Provider<ProjectManager> projectManagerProvider, ServerInitializer initializer) {
+        this.languages.addAll(languages);
+        this.launchers.putAll(languageServerLaunchers.stream()
+                        .collect(Collectors.toMap(LanguageServerLauncher::getLanguageId, launcher -> launcher)));
         this.projectManagerProvider = projectManagerProvider;
         this.initializer = initializer;
-        this.extensionToLauncher = new ConcurrentHashMap<>();
         this.projectToServer = new ConcurrentHashMap<>();
         this.initializer.addObserver(this);
-
-        for (LanguageServerLauncher launcher : languageServerLaunchers) {
-            for (String extension : launcher.getLanguageDescription().getFileExtensions()) {
-                extensionToLauncher.putIfAbsent(extension, new ArrayList<>());
-                extensionToLauncher.get(extension).add(launcher);
-            }
-        }
     }
 
     @Override
     public LanguageServer findServer(String fileUri) throws LanguageServerException {
         String path = URI.create(fileUri).getPath();
 
-        String extension = getFileExtension(path);
         String projectPath = extractProjectPath(path);
 
-        return findServer(extension, projectPath);
+        return doFindServer(projectPath, findLanguageId(path));
     }
 
-    @Nullable
-    protected LanguageServer findServer(String extension, String projectPath) throws LanguageServerException {
-        ProjectExtensionKey projectKey = createProjectKey(projectPath, extension);
-
-        for (LanguageServerLauncher launcher : extensionToLauncher.get(extension)) {
-            if (!projectToServer.containsKey(projectKey)) {
-                synchronized (launcher) {
-                    if (!projectToServer.containsKey(projectKey)) {
-                        LanguageServer server = initializer.initialize(launcher, projectPath);
-                        projectToServer.put(projectKey, server);
-                    }
+    private LanguageDescription findLanguageId(String path) {
+        for (LanguageDescription language : languages) {
+            if (matchesFilenames(language, path) || matchesExtensions(language, path)) {
+                return language;
+            }
+        }
+        return null;
+    }
+
+    private boolean matchesExtensions(LanguageDescription language, String path) {
+        if (language.getFileExtensions() != null) {
+            for (String extension : language.getFileExtensions()) {
+                if (path.endsWith(extension)) {
+                    return true;
                 }
             }
-            return projectToServer.get(projectKey);
         }
+        return false;
+    }
```

Comment:
It may be subjective but as JDK8 is out since a long time, I assume that ppl are fluent on using streams.
BTW if getFileExtensions would always return something (like empty) it could have been updated to only 
```
return language.getFileExtensions().stream().anyMatch(extension -> path.endsWith(extension));
```

which is definitely better.